File: packages/guards/src/phone-format-intl.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface PhoneFormatIntlOptions {
  action: 'block' | 'warn';
}

const PHONE_PATTERNS: RegExp[] = [
  /\+\d{1,3}\d{6,14}/,                        // E.164
  /\+\d{1,3}[\s.-]\d{1,4}[\s.-]\d{3,4}[\s.-]\d{3,4}/, // spaced intl
  /\(\d{2,4}\)\s?\d{3,4}[\s.-]?\d{3,4}/,      // (0XX) XXXX-XXXX
  /\b\d{3}[\s.-]\d{3}[\s.-]\d{4}\b/,          // US 123-456-7890
  /\b0\d{2,4}[\s.-]\d{6,8}\b/,                // UK/EU 0XX-XXXXXXXX
  /\b0\d{1,2}-\d{3,4}-\d{4}\b/,               // JP/KR 0X-XXXX-XXXX
];

export function phoneFormatIntl(options: PhoneFormatIntlOptions): Guard {
  return {
    name: 'phone-format-intl',
    version: '0.1.0',
    description: 'Detects international phone number formats',
    category: 'privacy',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const matched: string[] = [];

      for (const pattern of PHONE_PATTERNS) {
        const m = text.match(pattern);
        if (m) matched.push(m[0]);
      }

      const triggered = matched.length > 0;

      return {
        guardName: 'phone-format-intl',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score: triggered ? Math.min(matched.length / 2, 1.0) : 0,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { matchedCount: matched.length } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/phone_format_intl.py
"""Detect international phone number formats."""
from __future__ import annotations

import re
import time
from typing import Optional

from open_guardrail.core import GuardResult

_PHONE_PATTERNS: list[re.Pattern[str]] = [
    re.compile(r"\+\d{1,3}\d{6,14}"),
    re.compile(
        r"\+\d{1,3}[\s.-]\d{1,4}[\s.-]"
        r"\d{3,4}[\s.-]\d{3,4}"
    ),
    re.compile(r"\(\d{2,4}\)\s?\d{3,4}[\s.-]?\d{3,4}"),
    re.compile(r"\b\d{3}[\s.-]\d{3}[\s.-]\d{4}\b"),
    re.compile(r"\b0\d{2,4}[\s.-]\d{6,8}\b"),
    re.compile(r"\b0\d{1,2}-\d{3,4}-\d{4}\b"),
]


class _PhoneFormatIntl:
    def __init__(
        self, *, action: str = "block"
    ) -> None:
        self.name = "phone-format-intl"
        self.action = action

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        matched: list[str] = []

        for pat in _PHONE_PATTERNS:
            if pat.search(text):
                matched.append(pat.pattern)

        triggered = len(matched) > 0
        score = (
            min(len(matched) / 2, 1.0)
            if triggered
            else 0.0
        )
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="phone-format-intl",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "International phone number detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"matched_patterns": len(matched)}
                if triggered
                else None
            ),
        )


def phone_format_intl(
    *, action: str = "block"
) -> _PhoneFormatIntl:
    return _PhoneFormatIntl(action=action)
